src/server.c

```diff
@@ -5909,6 +5909,57 @@ void sendChildInfo(childInfoType info_type, size_t keys, char *pname) {
     sendChildInfoGeneric(info_type, keys, -1, pname);
 }
 
+/* Use 'MADV_DONTNEED' to release memory to operation system quickly. */
+void redisUnusedMemory(void *ptr) {
+    /* madvise(MADV_DONTNEED) may not work if enabled Transparent Huge Pages. */
+    if (server.thp_enabled == 1) return;
+    if (ptr == NULL) return;
+
+    size_t real_size = zmalloc_size(ptr);
+    if (real_size < server.page_size) return;
+
+    size_t page_size_mask = server.page_size - 1;
+    void *real_ptr = zmalloc_realptr(ptr);
+    char *aligned_ptr = (char *)((size_t)(real_ptr) & ~page_size_mask);
```

Comment:
You need to align the pointer upwards (find set of complete pages inside this allocation). 
This is probably what corrupted your tests
```suggestion
    char *aligned_ptr = (char *)((size_t)(real_ptr + page_size_mask) & ~page_size_mask);
```
You need to align the pointer upwards, to find a complete set of pages that fall inside the allocation (P. S. Comment needed). 
This is what probably corrupted your tests..

Reply:
truly thanks @oranagra I suddenly notice I should align the pointer upwards instead of downwards, because the memory address is increased upwards